balancer/weightedroundrobin/weightedroundrobin.go

```diff
@@ -27,3 +39,181 @@ const Name = "weighted_round_robin"
 type AddrInfo struct {
 	Weight uint32
 }
+
+type wrrBuilder struct{}
+
+// newBuilder creates a new roundrobin balancer builder.
+func newBuilder() balancer.Builder {
+	return &wrrBuilder{}
+}
+
+func init() {
+	balancer.Register(newBuilder())
+}
+
+func (*wrrBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return &wrrBalancer{
+		cc: cc,
+
+		subConns: make(map[string]*balancerAddrInfo),
+		scStates: make(map[balancer.SubConn]connectivity.State),
+		csEvltr:  &balancer.ConnectivityStateEvaluator{},
+	}
+}
+
+func (*wrrBuilder) Name() string {
+	return Name
+}
+
+type balancerAddrInfo struct {
+	SubConn balancer.SubConn
+	Weight  uint32
+}
+
+type wrrBalancer struct {
+	cc balancer.ClientConn
+
+	csEvltr *balancer.ConnectivityStateEvaluator
+	state   connectivity.State
+
+	subConns map[string]*balancerAddrInfo
+	scStates map[balancer.SubConn]connectivity.State
+	picker   balancer.Picker
+}
+
+func (*wrrBalancer) HandleSubConnStateChange(sc balancer.SubConn, state connectivity.State) {
+	panic("not implemented")
+}
+
+func (*wrrBalancer) HandleResolvedAddrs([]resolver.Address, error) {
+	panic("not implemented")
+}
+
+func (*wrrBalancer) Close() {
+}
+
+func (b *wrrBalancer) UpdateClientConnState(s balancer.ClientConnState) {
+	grpclog.Infoln("wrr: got new ClientConn state: ", s)
+	// addrWeights is the set converted from addrs, it's used for quick lookup of an address.
+	addrWeights := make(map[string]uint32)
+	for _, a := range s.ResolverState.Addresses {
+		weight := uint32(1)
+		if addrInfo, ok := a.Metadata.(*AddrInfo); ok {
+			weight = addrInfo.Weight
+		}
+		addrWeights[a.Addr] += weight
+		if _, ok := b.subConns[a.Addr]; !ok {
+			// a is a new address (not existing in b.subConns).
+			sc, err := b.cc.NewSubConn([]resolver.Address{a}, balancer.NewSubConnOptions{HealthCheckEnabled: true})
+			if err != nil {
+				grpclog.Warningf("base.baseBalancer: failed to create new SubConn: %v", err)
+				continue
+			}
+			balancerAddrInfo := balancerAddrInfo{SubConn: sc}
+			b.subConns[a.Addr] = &balancerAddrInfo
+			b.scStates[sc] = connectivity.Idle
+			sc.Connect()
+		}
+	}
+	for a, sc := range b.subConns {
+		if weight, ok := addrWeights[a]; !ok {
+			// a was removed by resolver.
+			b.cc.RemoveSubConn(sc.SubConn)
+			delete(b.subConns, a)
+			// Keep the state of this sc in b.scStates until sc's state becomes Shutdown.
+			// The entry will be deleted in HandleSubConnStateChange.
+		} else {
+			sc.Weight = weight
```

Comment:
When will the picker get the new weight?
I think this needs to update the picker as well.